File: TermiusTool.py
import os
import re
import subprocess
import sys
import requests
import winreg
import time

###################
# Node.js Functions #
###################

def check_nodejs():
    """
    Check if Node.js is installed
    Returns: bool
    """
    try:
        subprocess.run(['node', '--version'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        return True
    except FileNotFoundError:
        return False
    except Exception as e:
        print(f"Error checking Node.js: {e}")
        return False

def download_nodejs():
    """
    Download Node.js installer
    Returns: installer path
    """
    print('Downloading Node.js...')
    url = 'https://nodejs.org/dist/v22.11.0/node-v22.11.0-x64.msi'
    response = requests.get(url, stream=True)
    installer_path = os.path.join(os.environ['TEMP'], 'nodejs_installer.msi')
    
    total_size = int(response.headers.get('content-length', 0))
    block_size = 1024
    downloaded = 0
    
    with open(installer_path, 'wb') as f:
        for data in response.iter_content(block_size):
            downloaded += len(data)
            f.write(data)
            progress = downloaded * 100 / total_size
            done = int(50 * downloaded / total_size)
            sys.stdout.write('\r[{}{}] {:.1f}%'.format(
                '=' * done, ' ' * (50-done), progress
            ))
            sys.stdout.flush()
    print()
    
    return installer_path

def refresh_env():
    """Refresh current process environment variables"""
    try:
        # Get system environment variables
        with winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, r'SYSTEM\CurrentControlSet\Control\Session Manager\Environment', 0, winreg.KEY_READ) as key:
            path_value = winreg.QueryValueEx(key, 'Path')[0]
            
        # Get user environment variables
        with winreg.OpenKey(winreg.HKEY_CURRENT_USER, r'Environment', 0, winreg.KEY_READ) as key:
            user_path = winreg.QueryValueEx(key, 'Path')[0]
            
        # Merge system and user PATH
        full_path = f"{path_value};{user_path}"
        
        # Update current process environment variables
        os.environ['Path'] = full_path
        
        print("Environment variables refreshed")
        return True
    except Exception as e:
        print(f"Failed to refresh environment variables: {e}")
        return False

def install_nodejs():
    """
    Install Node.js
    Returns: (bool, str, str) - (success, node_path, npm_path)
    """
    try:
        # Ask user if they want to install Node.js
        install = input('Node.js is required. Install? (y/n): ').lower().strip() == 'y'
        if not install:
            return False, '', ''

        # Download and install
        installer_path = download_nodejs()
        print('Installing Node.js, please wait...')
        subprocess.run(['msiexec.exe', '/i', installer_path, '/quiet', '/norestart'], check=True)
        
        # Clean up installer
        os.remove(installer_path)
        
        # Refresh environment variables after installation
        if not refresh_env():
            print("Warning: Failed to refresh environment variables, may need to restart terminal")
        
        # Get Node.js installation path
        node_path = os.path.join(os.environ['ProgramFiles'], 'nodejs', 'node.exe')
        npm_path = os.path.join(os.environ['ProgramFiles'], 'nodejs', 'npm.cmd')
        
        if not os.path.exists(node_path) or not os.path.exists(npm_path):
            print('Node.js files not found, installation may have failed')
            return False, '', ''
            
        # Verify installation
        print('Verifying Node.js installation...')
        try:
            subprocess.run([node_path, '--version'], check=True, capture_output=True)
            print('Node.js installation verified!')
            return True, node_path, npm_path
        except subprocess.CalledProcessError:
            print('Node.js installation verification failed')
            return False, '', ''
        
    except Exception as e:
        print(f'Node.js installation failed: {str(e)}')
        return False, '', ''

def get_nodejs_path():
    """
    Get actual installation paths of Node.js and npm
    Returns: (str, str) - (node_path, npm_path)
    """
    try:
        # First, try using where command to find
        node_result = subprocess.run(['where', 'node'], 
                                   capture_output=True, 
                                   text=True)
        npm_result = subprocess.run(['where', 'npm.cmd'], 
                                  capture_output=True, 
                                  text=True)
        
        if node_result.returncode == 0 and npm_result.returncode == 0:
            node_path = node_result.stdout.splitlines()[0].strip()
            npm_path = npm_result.stdout.splitlines()[0].strip()
            # Verify found paths are usable
            if verify_nodejs_installation(node_path, npm_path):
                return node_path, npm_path
            
        # If where command fails, try looking up in registry
        try:
            with winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, r'SOFTWARE\Node.js', 0, winreg.KEY_READ | winreg.KEY_WOW64_64KEY) as key:
                install_dir = winreg.QueryValueEx(key, 'InstallPath')[0]
                node_path = os.path.join(install_dir, 'node.exe')
                npm_path = os.path.join(install_dir, 'npm.cmd')
                if os.path.exists(node_path) and os.path.exists(npm_path):
                    # Verify found paths are usable
                    if verify_nodejs_installation(node_path, npm_path):
                        return node_path, npm_path
        except WindowsError:
            pass
            
        # Finally, try default paths
        default_paths = [
            os.environ.get('ProgramFiles'),
            os.environ.get('ProgramFiles(x86)'),
            os.environ.get('LocalAppData')
        ]
        
        for base_path in default_paths:
            if base_path:
                node_path = os.path.join(base_path, 'nodejs', 'node.exe')
                npm_path = os.path.join(base_path, 'nodejs', 'npm.cmd')
                if os.path.exists(node_path) and os.path.exists(npm_path):
                    # Verify found paths are usable
                    if verify_nodejs_installation(node_path, npm_path):
                        return node_path, npm_path
                    
        print('No usable Node.js installation found, preparing to reinstall...')
        return None, None
        
    except Exception as e:
        print(f"Error finding Node.js path: {e}")
        return None, None

def verify_nodejs_installation(node_path, npm_path):
    """Verify if Node.js installation is working"""
    try:
        subprocess.run([node_path, '--version'], 
                      check=True, 
                      capture_output=True)
        subprocess.run([npm_path, '--version'], 
                      check=True, 
                      capture_output=True)
        return True
    except Exception:
        return False

##################
# ASAR Functions #
##################

def install_asar(npm_path):
    """
    Install asar globally
    Args: npm_path - Full path to npm executable
    Returns: bool
    """
    try:
        print(f'Installing asar using npm...')
        
        result = subprocess.run(
            [npm_path, 'install', '-g', 'asar'],
            shell=False,
            check=True,
            capture_output=True,
            text=True
        )
        
        print('asar installed successfully!')
        return True
        
    except subprocess.CalledProcessError as e:
        print(f'asar installation failed: {e.stderr if e.stderr else str(e)}')
        return False
    except Exception as e:
        print(f'asar installation failed: {str(e)}')
        return False

def main(disable_update=False):
    """Main function"""
    # 1. Check and install Node.js
    if not check_nodejs():
        success, node_path, npm_path = install_nodejs()
        if not success:
            print('Node.js not installed, exiting')
            sys.exit(1)
    else:
        node_path, npm_path = get_nodejs_path()
        if not node_path or not npm_path:
            success, node_path, npm_path = install_nodejs()
            if not success:
                print('Node.js installation failed, exiting')
                sys.exit(1)
        else:
            print('Node.js installed')

    # 2. Install asar
    print('Installing asar...')
    if not install_asar(npm_path):
        print('asar installation failed, exiting')
        sys.exit(1)
    
    print('Waiting for variable refresh...')
    time.sleep(5)
    refresh_env()
    
    # 3. Check files and processes
    asar_cmd = os.path.join(os.environ['APPDATA'], 'npm', 'asar.cmd')
    if not os.path.exists(asar_cmd):
        print('asar command not found')
        sys.exit(1)
    
    # Add auto-update prompt here, before file modifications
    disable_update = input('Disable auto-update? (y/n): ').lower().strip() == 'y'
    print('Auto-update will be ' + ('disabled' if disable_update else 'enabled'))
    
    base_path = os.path.join(os.path.expanduser("~"), 'AppData', 'Local', 'Programs', 'Termius', 'resources')
    asar_file = os.path.join(base_path, 'app.asar')
    extract_dir = os.path.join(base_path, 'app')

    print('Checking Termius process...')
    try:
        subprocess.run(['taskkill', '/F', '/IM', 'Termius.exe'], 
                      capture_output=True,
                      text=True)
        print('Termius closed, waiting for process exit...')
        time.sleep(2)
    except subprocess.CalledProcessError:
        print('Termius not running, continuing...')

    # 4. Extract and modify files
    if not os.path.exists(asar_file):
        print(f'Error: File not found {asar_file}')
        sys.exit(1)

    try:
        print('Extracting app.asar file...')
        result = subprocess.run([asar_cmd, 'extract', asar_file, extract_dir], 
                              capture_output=True, 
                              text=True,
                              check=True)
        print('app.asar extracted successfully!')
        
        # Delete app.asar file
        max_retries = 3
        retry_delay = 2
        for i in range(max_retries):
            try:
                if os.path.exists(asar_file):
                    os.remove(asar_file)
                    print('app.asar deleted')
                break
            except PermissionError:
                if i < max_retries - 1:
                    print(f'Delete failed, retrying... ({i+1}/{max_retries})')
                    time.sleep(retry_delay)
                else:
                    print('Warning: Cannot delete app.asar file, but crack will still work')
                    break
    except Exception as e:
        print(f'Error extracting or deleting file: {str(e)}')
        sys.exit(1)

    # 5. Disable auto update and complete modifications
    if disable_update:
        update_file = os.path.join(base_path, 'app-update.yml')
        if os.path.exists(update_file):
            os.remove(update_file)
            print('app-update.yml deleted, auto update disabled')
        else:
            print('app-update.yml not found')

    # Modify main program file
    assets_dir = os.path.join(extract_dir, 'background-process', 'assets')
    for root, dirs, files in os.walk(assets_dir):
        for file in files:
            if file.startswith('main-') and file.endswith('.js'):
                file_path = os.path.join(root, file)
                
                # Read file content
                with open(file_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                # Replace specified content
                old_code = "const e=await this.api.bulkAccount();"
                new_code = '''
var e=await this.api.bulkAccount();
e.account.pro_mode=true;
e.account.need_to_update_subscription=false;
e.account.current_period={
    "from": "2022-01-01T00:00:00",
    "until": "2099-01-01T00:00:00"
};
e.account.plan_type="Premium";
e.account.user_type="Premium";
e.student=null;
e.trial=null;
e.account.authorized_features.show_trial_section=false;
e.account.authorized_features.show_subscription_section=true;
e.account.authorized_features.show_github_account_section=false;
e.account.expired_screen_type=null;
e.personal_subscription={
    "now": new Date().toISOString().slice(0, -5),
    "status": "SUCCESS",
    "platform": "stripe",
    "current_period": {
        "from": "2022-01-01T00:00:00",
        "until": "2099-01-01T00:00:00"
    },
    "revokable": true,
    "refunded": false,
    "cancelable": true,
    "reactivatable": false,
    "currency": "usd",
    "created_at": "2022-01-01T00:00:00",
    "updated_at": new Date().toISOString().slice(0, -5),
    "valid_until": "2099-01-01T00:00:00",
    "auto_renew": true,
    "price": 12.0,
    "verbose_plan_name": "Termius Pro Monthly",
    "plan_type": "SINGLE",
    "is_expired": false
};
e.access_objects=[{
    "period": {
        "start": "2022-01-01T00:00:00",
        "end": "2099-01-01T00:00:00"
    },
    "title": "Pro"
}];
'''

                new_content = re.sub(re.escape(old_code), new_code, content)

                # Write back modified file
                with open(file_path, 'w', encoding='utf-8') as f:
                    f.write(new_content)

    print("Complete!")
    input("\nPress any key to exit...")

if __name__ == "__main__":
    main(disable_update=True)

File: TermiusTool.rs
use anyhow::{Context, Result};
use byteorder::{LittleEndian, ReadBytesExt};
use indicatif::{ProgressBar, ProgressStyle};
use regex::Regex;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::fs::{self, File};
use std::io::{self, BufReader, Cursor, Read, Seek, SeekFrom, Write};
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::thread;
use std::time::Duration;
use walkdir::WalkDir;

// Error type definitions
#[derive(thiserror::Error, Debug)]
enum TermiusToolError {
    #[error("File not found: {0}")]
    FileNotFound(String),
    
    #[error("Failed to extract ASAR file: {0}")]
    AsarExtractFailed(String),
    
    #[error("Failed to modify file: {0}")]
    FileModificationFailed(String),
    
    #[error("ASAR file format error: {0}")]
    AsarFormatError(String),
    
    #[error("IO error: {0}")]
    IoError(#[from] io::Error),
    
    #[error("JSON error: {0}")]
    JsonError(#[from] serde_json::Error),
}

// ASAR file processing structure
struct AsarFile {
    header: Value,
    header_size: u32,
    files: HashMap<String, Vec<u8>>,
}

impl AsarFile {
    // Load ASAR from file
    fn from_file(path: &Path) -> Result<Self> {
        let file = File::open(path)?;
        let mut reader = BufReader::new(file);
        
        // Read header size
        let header_size = reader.read_u32::<LittleEndian>()?;
        
        // Read header JSON
        let mut header_data = vec![0; header_size as usize];
        reader.read_exact(&mut header_data)?;
        
        let header: Value = serde_json::from_slice(&header_data)?;
        
        // Initialize file collection
        let mut files = HashMap::new();
        
        // Read all files
        Self::extract_files(&header["files"], "", &mut reader, header_size + 8, &mut files)?;
        
        Ok(Self {
            header,
            header_size,
            files,
        })
    }
    
    // Recursively extract files
    fn extract_files(
        node: &Value, 
        path: &str, 
        reader: &mut BufReader<File>, 
        base_offset: u32,
        files: &mut HashMap<String, Vec<u8>>
    ) -> Result<()> {
        if let Some(obj) = node.as_object() {
            for (name, info) in obj {
                let file_path = if path.is_empty() {
                    name.clone()
                } else {
                    format!("{}/{}", path, name)
                };
                
                if info["files"].is_object() {
                    // This is a directory
                    Self::extract_files(&info["files"], &file_path, reader, base_offset, files)?;
                } else if let (Some(size), Some(offset)) = (
                    info["size"].as_u64(),
                    info["offset"].as_u64()
                ) {
                    // This is a file
                    let size = size as usize;
                    let offset = offset as u64 + base_offset as u64;
                    
                    // Read file content
                    reader.seek(SeekFrom::Start(offset))?;
                    let mut content = vec![0; size];
                    reader.read_exact(&mut content)?;
                    
                    // Store file content
                    files.insert(file_path, content);
                }
            }
        }
        
        Ok(())
    }
    
    // Extract to directory
    fn extract_to_dir(&self, output_dir: &Path) -> Result<()> {
        fs::create_dir_all(output_dir)?;
        
        let pb = ProgressBar::new(self.files.len() as u64);
        pb.set_style(ProgressStyle::default_bar()
            .template("{spinner:.green} [{elapsed_precise}] [{bar:50.cyan/blue}] {pos}/{len} ({eta})")
            .unwrap()
            .progress_chars("#>-"));
        
        for (path, content) in &self.files {
            let full_path = output_dir.join(path);
            
            // Create parent directory
            if let Some(parent) = full_path.parent() {
                fs::create_dir_all(parent)?;
            }
            
            // Write file
            let mut file = File::create(&full_path)?;
            file.write_all(content)?;
            
            pb.inc(1);
        }
        
        pb.finish_with_message("Extraction complete");
        Ok(())
    }
}

// Close Termius process
fn kill_termius() -> Result<()> {
    println!("Checking Termius process...");
    let _ = Command::new("taskkill")
        .args(&["/F", "/IM", "Termius.exe"])
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status();
    
    println!("Waiting for process to exit...");
    thread::sleep(Duration::from_secs(2));
    Ok(())
}

// Modify JS file
fn modify_js_file(extract_dir: &Path) -> Result<bool> {
    println!("Modifying files...");
    let mut modified = false;
    
    // Find and modify main JS file
    let assets_dir = extract_dir.join("background-process").join("assets");
    
    for entry in WalkDir::new(&assets_dir).into_iter().filter_map(|e| e.ok()) {
        let path = entry.path();
        if path.is_file() && path.extension().map_or(false, |ext| ext == "js") {
            let file_name = path.file_name().unwrap().to_string_lossy();
            if file_name.starts_with("main-") {
                // Read file content
                let content = fs::read_to_string(path)?;
                
                // Find code to replace
                let pattern = r"const e=await this\.api\.bulkAccount\(\);";
                let re = Regex::new(pattern)?;
                
                if re.is_match(&content) {
                    // Replace code
                    let replacement = r#"
var e=await this.api.bulkAccount();
e.account.pro_mode=true;
e.account.need_to_update_subscription=false;
e.account.current_period={
    "from": "2022-01-01T00:00:00",
    "until": "2099-01-01T00:00:00"
};
e.account.plan_type="Premium";
e.account.user_type="Premium";
e.student=null;
e.trial=null;
e.account.authorized_features.show_trial_section=false;
e.account.authorized_features.show_subscription_section=true;
e.account.authorized_features.show_github_account_section=false;
e.account.expired_screen_type=null;
e.personal_subscription={
    "now": new Date().toISOString().slice(0, -5),
    "status": "SUCCESS",
    "platform": "stripe",
    "current_period": {
        "from": "2022-01-01T00:00:00",
        "until": "2099-01-01T00:00:00"
    },
    "revokable": true,
    "refunded": false,
    "cancelable": true,
    "reactivatable": false,
    "currency": "usd",
    "created_at": "2022-01-01T00:00:00",
    "updated_at": new Date().toISOString().slice(0, -5),
    "valid_until": "2099-01-01T00:00:00",
    "auto_renew": true,
    "price": 12.0,
    "verbose_plan_name": "Termius Pro Monthly",
    "plan_type": "SINGLE",
    "is_expired": false
};
e.access_objects=[{
    "period": {
        "start": "2022-01-01T00:00:00",
        "end": "2099-01-01T00:00:00"
    },
    "title": "Pro"
}];
"#;
                    
                    let new_content = re.replace(&content, replacement).to_string();
                    fs::write(path, new_content)?;
                    
                    println!("Modified file: {}", file_name);
                    modified = true;
                    break;
                }
            }
        }
    }
    
    if !modified {
        println!("Warning: Required file not found, Termius version might have been updated, please download again");
        println!("Feature configuration might not work properly");
    }
    
    Ok(modified)
}

// Disable auto update
fn disable_auto_update(base_path: &Path) -> Result<bool> {
    let update_file = base_path.join("app-update.yml");
    if update_file.exists() {
        fs::remove_file(&update_file)?;
        println!("Deleted app-update.yml, auto update disabled");
        Ok(true)
    } else {
        println!("app-update.yml not found");
        Ok(false)
    }
}

// Main function
fn main() -> Result<()> {
    println!("Termius Pro Configuration Tool");
    println!("====================================");
    
    // Ask about auto update
    println!("Disable auto-update? (y/n): ");
    let mut input = String::new();
    io::stdin().read_line(&mut input)?;
    let disable_update = input.trim().eq_ignore_ascii_case("y");
    
    println!("Auto-update will be {}", if disable_update { "disabled" } else { "enabled" });
    
    // Set paths
    let home = std::env::var("USERPROFILE").context("Failed to get user home directory")?;
    let base_path = PathBuf::from(&home)
        .join("AppData")
        .join("Local")
        .join("Programs")
        .join("Termius")
        .join("resources");
    
    let asar_file = base_path.join("app.asar");
    let extract_dir = base_path.join("app");
    
    // Check if Termius is installed
    if !asar_file.exists() {
        return Err(TermiusToolError::FileNotFound(format!("File not found {}", asar_file.display())).into());
    }
    
    // Close Termius process
    kill_termius()?;
    
    // Backup original file
    let backup_file = asar_file.with_extension("asar.backup");
    if !backup_file.exists() {
        println!("Creating backup: {}", backup_file.display());
        fs::copy(&asar_file, &backup_file)?;
    }
    
    // Extract asar file
    println!("Extracting app.asar file...");
    if extract_dir.exists() {
        fs::remove_dir_all(&extract_dir)?;
    }
    
    // Use pure Rust to extract asar file
    let asar = AsarFile::from_file(&asar_file)
        .context("Failed to parse asar file")?;
    
    asar.extract_to_dir(&extract_dir)
        .context("Failed to extract asar file contents")?;
    
    println!("app.asar extracted successfully!");
    
    // Modify JS file
    modify_js_file(&extract_dir)?;
    
    // Disable auto update
    if disable_update {
        disable_auto_update(&base_path)?;
    }
    
    println!("\nComplete!");
    println!("Termius Pro has been configured");
    println!("Press Enter to exit...");
    let mut input = String::new();
    io::stdin().read_line(&mut input)?;
    
    Ok(())
}